File: adventofcode/2019/d1/d1.py
masses = []
with open('./input.txt') as f:
    for i, line in enumerate(f):
        masses.append(int(line))


def calc_fuel(mass):
    return mass//3 - 2


def calc_true_fuel(mass):
    fuel = calc_fuel(mass)
    return 0 if fuel <= 0 else fuel + calc_true_fuel(fuel)


# p1
print(sum(map(calc_fuel, masses)))

# p2
print(sum(map(calc_true_fuel, masses)))

File: adventofcode/2019/d1/d1.rs
use std::cmp::max;

fn parse_input() -> Vec<i32> {
  let file = include_str!("./input.txt");

  file.lines()
    .map(|x| i32::from_str_radix(x, 10).unwrap())
    .collect()
}

fn calc_fuel(mass:&i32) -> i32 {
  max(mass / 3 - 2, 0)
}

fn calc_true_fuel(mass:&i32) -> i32 {
  match calc_fuel(mass) {
    0 => 0,
    fuel => fuel + calc_true_fuel(&fuel),
  }
}

fn main() {
  let masses = parse_input();

  // p1
  println!("{}", masses.iter().map(calc_fuel).sum::<i32>());

  // p2
  println!("{}", masses.iter().map(calc_true_fuel).sum::<i32>());
}
